_posts/2017/2017-03-10-creating-a-sales-analytics-app-with-ruby-on-rails-2.md

```diff
@@ -0,0 +1,84 @@
+This way we could add encryption easily where needed. But then Devise said **no**.
+As it makes sense on an email field, the library turns that field into lowercase and as we where encrypting the email field and overwriting it, it turned `randomEncryptedStringForEmail` into `randomencryptedstringforemail`, which is not the same when decrypting it.
+
+After some hours of not understanding anything, I realized the lowercase issue and opted to create another field on the database for the merchant to store the encrypted value, so Devise couldn't lowercase it.
+With that, we had to change it everywhere, backend code and UI code, but to be honest, with `find and replace` happened to be an easy task.
+
+It wasn't until later when we saw the beast we created.
+The email field was used for **everything** (Both in Devise and in our App) that had to do with the User.
+Not only to update the email, update the password or restore password, but all over the core.
+As this kind of platforms usually have the User model in the middle, because you need to display all the data regarding to one specific user, we basically rewrote the all or at least most of the sessions and password code.
+That affected also to other parts of the Devise library as `reset_tokens` or `last_update` dates, which required even more core rewriting.
+
+## Hosting the app
+
+Now that we have dealt with all the issues on the app, here comes the hosting debate.
+Our team has experience on both [Heroku](https://www.heroku.com) and [AWS](https://aws.amazon.com) deployment so those are the main contenders.
+
+Following the philosophy on fast delivery, we started with Heroku, which allows us to deploy on one console command and to add some useful plugins like [Papertrail](https://papertrailapp.com) to store and search the logfiles or [New Relic](https://newrelic.com/) to check the application performance, something really critical to decide the host.
+
+If we end up with lots of traffic and big RAM usage, we may change to AWS so we can have a machine more suited to our needs.
+The downside of this scenario is that we will have to set up everything by ourselves, the machine, the deployment process and so on.
+Lucky for us, we have [Capistrano](http://capistranorb.com/), a tool written on Ruby (but available for other languages also) for remote server automation and deployment.
+
+[Docker](https://www.docker.com) is another option for us on the deployment side of the things to do when not choosing Heroku.
+We also have experience on it and in the case we needed support, our colleagues on ePages are [experienced](https://developer.epages.com/blog/2016/07/05/containerdays-hamburg.html) on both [Docker](https://developer.epages.com/blog/2016/11/28/xp-days-germany-2016.html) and [Kubernetes](https://developer.epages.com/blog/2016/08/09/how-to-setup-a-ha-kubernetes-cluster-etcd-cluster-with-ssl.html), so we know that we are more than covered on those subjects.
```

Comment:
... our colleagues at ePages are experienced on [Docker](https://developer.epages.com/blog/2016/07/05/containerdays-hamburg.html) and [Kubernetes](https://developer.epages.com/blog/2016/08/09/how-to-setup-a-ha-kubernetes-cluster-etcd-cluster-with-ssl.html). So we know, that we are more than covered on those subjects.